Semantic text dense vector support

carlosdelest opened the pull request:
Adds support for dense vector models.

To make sure the field mappers receive enough information to perform indexing, model information is included into each document that is ingested under `model_settings` field:

```json
{
    "infer_field": "these are not the droids you're looking for. He's free to go around",
    "non_infer_field": "that's no moon. It's a space station",
    "_semantic_text_inference": {
      "infer_field": {
        "model_settings": {
          "inference_id": "my_e5_small",
          "task_type": "text_embedding",
          "dimensions": 384
        },
        "inference_results": [
          {
            "inference": [
              0.006414609029889107,
              0.00162538792937994,
              -0.05126339942216873,
              -0.05343875288963318,
              0.009770095348358154,
              0.03464227914810181
            ],
            "text": "these are not the droids you're looking for. He's free to go around"
          }
        ]
      }
    }
  }
```

Commit:
Move SimilarityMeasure to server code

Commit:
Add dimensions and similarity to ServiceSettings, create ModelSettings class

Commit:
Change implementation of asMap() to avoid extra nesting in inference results

Commit:
Change inference results structure

Commit:
Field mapper uses new inference results structure

Commit:
Fix BulkOperationTests

Commit:
Fix tests of SemanticTextInferenceResultFieldMapper

Commit:
Revert "Change implementation of asMap() to avoid extra nesting in inference results"

This reverts commit bd4e19e13f3ad6d54d9a2c38d34562a91b47ac8d.

Commit:
Use coordination format instead of changing the asMap() implementation

Commit:
Fix tests for new results structure

Commit:
Add service extension for dense vector embeddings

Commit:
Add tests for dense vector embeddings

Commit:
Fix bug in model settings

Commit:
Initial work on inference field results mapping for validation

Commit:
Fix spotless

Commit:
Refactored inference services with common abstract class

Commit:
Merge branch 'feature/semantic-text' into carlosdelest/semantic-text-dense-vector-support

# Conflicts:
#	server/src/main/java/org/elasticsearch/inference/ServiceSettings.java

carlosdelest commented on server/src/main/java/org/elasticsearch/action/bulk/BulkShardRequestInferenceProvider.java:
```diff
@@ -90,7 +92,13 @@ public void onResponse(ModelRegistry.UnparsedModel unparsedModel) {
                             InferenceProvider inferenceProvider = new InferenceProvider(
-                                service.get().parsePersistedConfig(inferenceId, unparsedModel.taskType(), unparsedModel.settings()),
+                                service.get()
+                                    .parsePersistedConfigWithSecrets(
```
Secrets are needed in order to perform inference on external services

carlosdelest commented on server/src/main/java/org/elasticsearch/action/bulk/BulkShardRequestInferenceProvider.java:
```diff
@@ -259,35 +267,22 @@ public void onResponse(InferenceServiceResults results) {
+                        for (InferenceResults inferenceResults : results.transformToCoordinationFormat()) {
+                            String inferenceFieldName = inferenceFieldNames.get(i++);
+                            Map<String, Object> inferenceFieldResult = new LinkedHashMap<>();
+                            inferenceFieldResult.putAll(new ModelSettings(inferenceProvider.model).asMap());
```
Add model settings information to make it available to field mapping

carlosdelest commented on server/src/test/java/org/elasticsearch/action/bulk/BulkOperationTests.java:
Some refactorings to provide mocks for the new methods used to retrieve config and inference

Commit:
Add javadoc

Commit:
Merge branch 'feature/semantic-text' into carlosdelest/semantic-text-dense-vector-support

# Conflicts:
#	x-pack/plugin/inference/qa/test-service-plugin/src/main/java/org/elasticsearch/xpack/inference/mock/AbstractTestInferenceService.java
#	x-pack/plugin/inference/qa/test-service-plugin/src/main/java/org/elasticsearch/xpack/inference/mock/TestDenseInferenceServiceExtension.java
#	x-pack/plugin/inference/qa/test-service-plugin/src/main/java/org/elasticsearch/xpack/inference/mock/TestSparseInferenceServiceExtension.java

Commit:
Fixing tests after merge

Commit:
Fix tests

Commit:
Change back javadoc

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

Commit:
Merge branch 'feature/semantic-text' into carlosdelest/semantic-text-dense-vector-support

ldematte committed:
Fix node_selector in new esql test (#105943)

tvernum committed:
Fix gradle run on Serverless (#105938)

On Serverless it is not possible to configure deprecation indexing (it
is always off). This commit updates the behaviour of
`ElasticsearchCluster` to no longer attempt to configure deprecation
indexing on stateless nodes.

ldematte committed:
YAML test framework: re-introduce `requires` section and `cluster_features` conditions (#105763)

martijnvg committed:
Unmute testRollupNonTSIndex() and (#105949)

increase ILM logging to TRACE.

Relates to #103981

benwtrent committed:
Test mute for #105952 (#105953)

mute for https://github.com/elastic/elasticsearch/issues/105952

kderusso commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextInferenceResultFieldMapper.java:
```diff
@@ -239,29 +242,51 @@ private static void parseFieldInferenceResultElement(
+                }
+            }
+            Integer dimensions = modelSettings.dimensions();
+            if (dimensions == null) {
```
Couldn't dims be calculated automatically if they weren't provided?

carlosdelest replied:
They could! But it would signal that we're not getting what we expect from the inference results, which will always include them for dense vectors.

kderusso reviewed (commented):
Nice work!

benwtrent committed:
Add note about optional times and epochs (#105786)

arteam committed:
[test] Disable index.shard.check_on_startup for searchable snapshot tests (#105731)



This setting requires expensive processing due to verification the integrity of many important files during a shard recovery or relocation.

Therefore, it takes lots of time for the files to clean up and the assertShardFolder check may not complete in 30s.

Fixes #105202

DaveCTurner committed:
Fix TransportSLMGetExpiredSnapshotsActionTests (#105950)

Addresses test bug introduced in #105721: we must consume all the
`SnapshotInfo` instances before completing the final listener.

Closes #105922

carlosdelest commented on server/src/main/java/org/elasticsearch/inference/ModelSettings.java:
```diff
@@ -0,0 +1,100 @@
+    private final SimilarityMeasure similarity;
+
+    public ModelSettings(TaskType taskType, String inferenceId, Integer dimensions, SimilarityMeasure similarity) {
+        Objects.requireNonNull(taskType, "task type must not be null");
```
`TaskType` is an enum, so there's no need for that

kderusso replied:
☕

carlosdelest commented on x-pack/plugin/inference/src/yamlRestTest/java/org/elasticsearch/xpack/inference/InferenceRestIT.java:
```diff
@@ -21,7 +21,7 @@ public class InferenceRestIT extends ESClientYamlSuiteTestCase {
         .setting("xpack.security.enabled", "false")
         .setting("xpack.security.http.ssl.enabled", "false")
-        .plugin("org.elasticsearch.xpack.inference.mock.TestInferenceServicePlugin")
+        .plugin("inference-service-test")
```
TIL as well 😆

carlosdelest commented on x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/10_semantic_text_inference.yml:
```diff
@@ -46,24 +76,50 @@ setup:
 
-    - match: { _source._semantic_text_inference.inference_field.0.text: "inference test" }
-    - match: { _source._semantic_text_inference.another_inference_field.0.text: "another inference test" }
+    - match: { _source._semantic_text_inference.inference_field.inference_results.0.text: "inference test" }
```
there's already a skip on setup for versions < 8.12.99. Do you think this can fail for other reasons?

kderusso replied:
Ah that's my bad, it didn't show up in the diff.

jimczi commented on x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapper.yml:
```diff
@@ -0,0 +1,153 @@
+                  - text: "inference test"
+                    inference: [0.1, 0.2, 0.3, 0.4, 0.5]
+                  - text: "another inference test"
+                    inference: [-0.1, -0.2, -0.3, -0.4, -0.5]
```
We should check that the mapping is as expected?

carlosdelest replied:
We won't be doing mapping updates, just the raw indexing into the Lucene doc. How can we check that?

carlosdelest replied:
I've added some tests for checking the mapping does not create additional fields, is that what you were thinking of?

jimczi replied:
I was thinking of checking that the field is updated with the right parameters (dimensions and similarity).

carlosdelest replied:
It is not - there is no change to the mappings that reflect this change:
-  The semantic_text field keeps having just `inference_id` and `type`.
- There are no other fields created, as the metadata field mapper deals with the results without creating mapping changes.

Is there any mapping change you'd expect from this operation?

jimczi replied:
oh I see, the mapping change is on the inner field and they are not exposed in the mapping?

carlosdelest replied:
That is correct. There is no mapping change, the metadata field mapper performs the indexing directly, without a mapping change.

AFAIU we can't do updates in the metadata field mapper for holding the changes. And we can't create a mapping with the same field path as a metadata field.

Is there something I'm missing?

carlosdelest replied:
There's some unit tests provided by @Mikep86 that check the Lucene doc structure.

Also, we can check the end-to-end scenario once we have the semantic_query ready, so we can check retrieving docs work

jimczi commented on x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/20_semantic_text_field_mapper.yml:
```diff
@@ -0,0 +1,153 @@
+                      feature_1: 0.1
+                      feature_2: 0.2
+                      feature_3: 0.3
+                      feature_4: 0.4
```
We should check the response too? And then verify that the mapping is as expected?

jimczi replied:
I misread, I thought it was a bulk index but it's just one document sorry.

jimczi committed:
Fix performance bug in `SourceConfirmedTextQuery#matches` (#105930)

This change ensures that the matches implementation of the `SourceConfirmedTextQuery` only checks the current document instead of calling advance on the two phase iterator. The latter tries to find the first doc that matches the query instead of restricting the search to the current doc. This can lead to abnormally slow highlighting if the query is very restrictive and the highlight is done on a non-matching document.

Closes #103298

kderusso committed:
Bugfix for mixed version cluster queries using text expansion (#105912)

* Bugfix for CCR queries using text expansion

* Fix test

* PR feedback

* Fix test

* Minor cleanup

* Edit comment

* One more comment clarification

---------

Co-authored-by: Elastic Machine <elasticmachine@users.noreply.github.com>

Commit:
Merge branch 'main' into carlosdelest/semantic-text-dense-vector-support

Commit:
Remove duplicate class

Commit:
Check mappings are as expected

Commit:
I hate YAML

jimczi reviewed (approved):
I left a minor comment, LGTM otherwise

benwtrent reviewed (approved):
Good step in the right direction. I like the reuse of the mapper builders. Maybe even allowing MORE model types in the future. 

We do need to figure out how to prevent inference results formats from changing once we have parsed a particular kind of field, but that discussion is outside the scope of this pr.
